Thunderbolt fixes

westeri opened the pull request:
This pull request contains fixes for review comments after the Titan Ridge support patch series was merged.

There is one additional commit that makes the Thunderbolt plugin to parse nvm_version everywhere in the same way.

westeri commented:
@YehezkelShB, these should fix your findings. Could you please review?

hughsie commented on plugins/thunderbolt/fu-thunderbolt-image.c:
```diff
@@ -491,24 +491,77 @@ get_host_locations (guint16 id)
+	if (image_mc_loc.offset == 0) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE,
+				     "The FW image file misses multi controller entry");
```
drop the `The` -- perhaps something like `firmware does not have multi-controller entry point`

hughsie commented on plugins/thunderbolt/fu-thunderbolt-image.c:
```diff
@@ -491,24 +491,77 @@ get_host_locations (guint16 id)
+	if (fw_mc_loc.len != image_mc_loc.len) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE,
+				     "The FW image multi controller entry length mismatch");
```
same

hughsie commented on plugins/thunderbolt/fu-thunderbolt-image.c:
```diff
@@ -491,24 +491,77 @@ get_host_locations (guint16 id)
+	if (memcmp (fw_mc->data, image_mc->data, fw_mc->len) != 0) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE,
+				     "The FW image multi controller entry mismatch");
```
same

hughsie commented on plugins/thunderbolt/fu-thunderbolt-image.c:
```diff
@@ -520,11 +573,9 @@ get_device_locations (guint16 id, const FuThunderboltFwObject *fw, GError **erro
-		if (read_drom_entry_location (fw, DROM_ENTRY_MC, &TR[1],
-					      &error_local))
+		 * compare it against the image first. */
+		if (compare_device_mc (fw, image, error))
```
You need the error_local and g_deubg perhaps? Either you return VALIDATION_FAILED and set error, or return TR and don't touch error. If that's what's happening, it's slightly odd to rely on the fallthrough IMHO.

westeri replied:
Hmm, so the idea is to return VALIDATION_FAILED and propagate error if the `compare_device_mc ()` fails. Otherwise return TR. I wonder if this would be better than the fallthrough case:
```
if (compare_device_mc (fw, image, error))
        return TR;
return NULL;
```
?

westeri commented:
Looks like self tests are using "raw" nvm_version such as "20.0" and the patch included in this pull request makes it fail because it converts all nvm_version strings to the fwupd format "20.00". I suppose here the correct fix is to update the self tests accordingly? @superm1?

YehezkelShB commented on plugins/thunderbolt/fu-thunderbolt-image.c:
```diff
@@ -491,24 +491,77 @@ get_host_locations (guint16 id)
+		   GError **error)
+{
+	FuThunderboltFwLocation image_mc_loc = { .section = DROM_SECTION, .description = "Multi Controller" };
+	FuThunderboltFwLocation fw_mc_loc = { .section = DROM_SECTION, .description = "Multi Controller" };
```
Matter of taste, but some would prefer `= image_mc_loc` instead of repeating the literal info.

YehezkelShB replied:
Also, I think most of the code uses `controller`, as `fw` is a bit confusing (it can be the FW image).

YehezkelShB replied:
(It makes sense to change the parameter name of `get_device_locations()` too as now it gets two different `FuThunderboltFwObject` parameters)

westeri commented:
I updated the branch according to your comments. I also updated self tests to include `nvm_parsed_version` that should match the version returned by fwupd functions.

hughsie commented on plugins/thunderbolt/fu-plugin-thunderbolt.c:
```diff
@@ -174,6 +174,20 @@ fu_plugin_thunderbolt_find_nvmem (GUdevDevice  *udevice,
+				(guint) g_ascii_strtoull (split[1], NULL, 16));
+}
+
 static const gchar *
```
this can't be const if it's returning an allocated string

YehezkelShB replied:
@hughsie
"Can't" technically or by convention?

westeri commented:
Hopefully fixed the build issue now.

superm1 commented on plugins/thunderbolt/fu-self-test.c:
```diff
@@ -1174,7 +1180,7 @@ test_update_fail (ThunderboltTest *tt, gconstpointer user_data)
 	/* version should *not* have changed */
```
I think this comment should actually be updated since the version is changing.

superm1 commented on plugins/thunderbolt/fu-self-test.c:
```diff
@@ -792,18 +793,21 @@ static MockDevice root_one = {
 	.name = "Laptop",
 	.id = "0x23",
 	.nvm_version = "20.0",
+	.nvm_parsed_version = "20.00",
```
With this change, I wonder if it's also adding a device that changes in potentially non-obvious way to self tests since it's not obvious immediately that's what's happening.
eg `.nvm_version = 2.2` -> `.nvm_parsed_version = 02.02`

westeri replied:
Not sure I follow. `nvm_parsed_version` is the expected version string after fwupd has added the corresponding device. It uses the format `%02X.%02X` so 2.2 would be 02.02. But that's exactly what happens with real devices, no?

superm1 replied:
Yes, but it's not an obvious behavior.  At least when you're given a string `X.Y` but `X` and `Y` are one digit usually you treat it as `0X.0Y`.  So when the interface from the kernel gives `23.5` unless you've been watching closely you won't know that's major 23 and minor 5 and might read it as `23.50`.

I just want to make sure that nuance is captured in the self tests as the people who aren't us who look at this code down the road might not realize that's what happening.

So I think either adding another device to self tests or changing one of them to have a non-zero minor would be good.

westeri replied:
OK, understood now.

superm1 commented on plugins/thunderbolt/fu-plugin-thunderbolt.c:
```diff
@@ -413,7 +411,7 @@ static void
 {
 	FuDevice *dev;
-	const gchar *version;
+	g_autofree gchar *version;
```
missing `= NULL`

westeri replied:
Oops. Not used to these g_auto* things yet.

I don't see these warnings when I run:
```
ninja -C build
```
Am I missing something?

superm1 replied:
Looks like CI is showing them failing too.  CI runs with `-Werror`, see if they show up from that.

superm1 commented on plugins/thunderbolt/fu-plugin-thunderbolt.c:
```diff
@@ -616,7 +614,7 @@ on_wait_for_device_added (FuPlugin    *plugin,
 	const gchar *uuid;
 	const gchar *path;
-	const gchar *version;
+	g_autofree gchar *version;
```
missing `= NULL;`

westeri commented:
I fixed the issues CI reported and changed version string of one device to be 20.2 which parses to 20.02.

superm1 commented:
I didn't test it this particular PR yet but I did test Mika's earlier one that already merged.

YehezkelShB commented:
(But I have no way to test it. Sorry.)
